--rm delete container when create failed

azurezk opened the pull request:
Signed-off-by: Zhang Kun zkazure@gmail.com

fixes #12444 
fixes #6757
When daemon.newContainer(name, config, imgID) success, return the container instead of nil so client can deleted it  if following creating procedure failed

If following container creating procedure failed, the container id will add in the error info.

calavera commented:
I'd rather return a proper json with the container id than adding the id to the error, tbh.

azurezk commented:
Hi , @calavera , If server return a proper json, the http status would be 200?  How to distinguish a successful return and a create failure?

jfrazelle commented:
Do we really need this? can you explain a use case. I understand --rm on run but kinda weird on create tbh

icecrime commented:
This fixes a seemingly legitimate bug, but I agree the solution is not appropriate, if only because every API clients to do the same string manipulation to extract the container ID. Also, you can return a JSON with a non 200 status.

tonistiigi commented:
We should evaluate if this could be solved in the daemon side. If I understand this correctly then `POST /containers/create` returns error  but container is still created. I think this shouldn't happen and if error in create code happens after container already exists server should just clean up after itself.

azurezk commented:
@tonistiigi, solved in the daemon side means clean up when create failure no mater there is a --rm flag or not?

tonistiigi commented:
@azurezk yes, IMHO if the container creation failed it shouldn't have side effects

azurezk commented:
OK, trying to do this

azurezk committed:
fix in daemon side

Signed-off-by: Zhang Kun <zkazure@gmail.com>

goformat

Signed-off-by: Zhang Kun <zkazure@gmail.com>

fix small

Signed-off-by: Zhang Kun <zkazure@gmail.com>

change to rm

Signed-off-by: Zhang Kun <zkazure@gmail.com>

handler other error

Signed-off-by: Zhang Kun <zkazure@gmail.com>

unique ERR

Signed-off-by: Zhang Kun <zkazure@gmail.com>

setHostConfig_fail_test

Signed-off-by: Zhang Kun <zkazure@gmail.com>

format

Signed-off-by: Zhang Kun <zkazure@gmail.com>

err handle and modify test

Signed-off-by: Zhang Kun <zkazure@gmail.com>

golint error

Signed-off-by: Zhang Kun <zkazure@gmail.com>

azurezk commented:
have fixed in daemon side, please review :)  @tonistiigi

azurezk commented:
@jfrazelle, I think if a container is created failed, it is expected that it doesn't exist and you can create another container with the same name. But it will cause conflict if not cleaned up. #6757 and #12444  have discuss this

tonistiigi commented:
I think we don't need a separate `Cleanup` function, partly because `daemon.Cleanup` should probably do something else. Calling just `daemon.rm()` should be enough there.

We should try to catch the other errors as well. You could use the following pattern after container is created(watch out to use single err variable though). 

```
defer func() {
  if err != nil  {
    daemon.rm()
  }
}
```

This needs integration tests.

azurezk commented:
have updated

tonistiigi commented:
@azurezk   This shouldn't work (and doesn't work for me) because the later errors are scoped and have their own error variable. We need to only have a single err variable to make this work. This also needs `integration-cli` tests because it's kind of tricky.

calavera commented:
> We need to only have a single err variable to make this work.

try this trick, you don't really need a single err variable in the method:

```go
func (daemon *Daemon) Create(config *runconfig.Config, hostConfig *runconfig.HostConfig, name string) (c *Container, s []string, retError error) {
    defer func() {
      if retError != nil  {
        daemon.rm()
      }
    }
}
```

The returning error is always associated with the named `retError` in the method definition.

calavera commented:
look at this example in the go playground: https://play.golang.org/p/SqSA-ChGOp

azurezk commented:
Thanks for help!  :) @calavera , @tonistiigi , integration-cli has passed

tonistiigi commented:
@azurezk Great. This looks good now but we need to add a new `integration-cli` test that makes a failed create call and tests that the container was cleaned up properly before we can merge. Let us know if you need help with that or if you want us to take over and make the test for you.

calavera commented:
Indeed, this change needs new integration tests.

azurezk commented:
I add a test to make it failed at "daemon.setHostConfig(container, hostConfig)", but there are many other processes I don't know how to make them failed. Please help or feel free to take over. @tonistiigi  @calavera

tonistiigi commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2816,3 +2816,21 @@ func (s *DockerSuite) TestRunCapAddSYSTIME(c *check.C) {
 }
+
+// run create container failed should remove container
+func (s *DockerSuite) TestRunCreateContainerFaliedCleanUp(c *check.C) {
```
s/Falied/Failed/

tonistiigi commented on daemon/create.go:
```diff
@@ -75,6 +75,12 @@ func (daemon *Daemon) Create(config *runconfig.Config, hostConfig *runconfig.Hos
 	}
+	defer func() {
+		if retErr != nil {
+			daemon.rm(container, false)
```
I think we could log out if `rm()` returns an error here.

tonistiigi commented:
@azurezk Don't worry about the other cases if they aren't really reproducible from client. Its important to test the invalid link thing that users are hitting.

Small comments, overall looks good to me.

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2816,3 +2816,21 @@ func (s *DockerSuite) TestRunCapAddSYSTIME(c *check.C) {
+	}
+
+	out, err = getAllContainers()
+	if err != nil {
```
`c.Assert(err, check.IsNil, check.Commentf(out))`

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2816,3 +2816,21 @@ func (s *DockerSuite) TestRunCapAddSYSTIME(c *check.C) {
+func (s *DockerSuite) TestRunCreateContainerFaliedCleanUp(c *check.C) {
+	name := "flowers"
+	out, _, err := dockerCmdWithError("run", "--name", name, "--link", "nothing:nothing", "busybox")
+	if err == nil {
```
c.Assert(err, check.Not(check.IsNil))

cpuguy83 commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2816,3 +2816,21 @@ func (s *DockerSuite) TestRunCapAddSYSTIME(c *check.C) {
+		c.Fatal(out, err)
+	}
+
+	if out != "" {
```
c.Assert(out, check.Equals, "")

But also it may be better to instead just call `inspect

cpuguy83 commented on daemon/create.go:
```diff
@@ -40,7 +40,7 @@ func (daemon *Daemon) ContainerCreate(name string, config *runconfig.Config, hos
 
 // Create creates a new container from the given configuration with a given name.
-func (daemon *Daemon) Create(config *runconfig.Config, hostConfig *runconfig.HostConfig, name string) (*Container, []string, error) {
+func (daemon *Daemon) Create(config *runconfig.Config, hostConfig *runconfig.HostConfig, name string) (retC *Container, retS []string, retErr error) {
```
Can we not use a named return and instead initialize `err` for the defer to use?

calavera replied:
we're using names returns to capture the error in the defer. I rather do that than trying to make sure the err in the defer is the correct one, this method is rather big for that.

cpuguy83 replied:
ok, works for me.

azurezk commented:
have updated :)  add daemon rm() error handle @tonistiigi  and inspect the container in test @cpuguy83

tonistiigi commented:
LGTM

cpuguy83 commented:
Looking at this again... shouldn't this be handled on the client since `--rm` is a client-side feature?

tonistiigi commented:
@cpuguy83 This doesn't actually have anything to do with `--rm` (although the original issue describes it like that). When the request to `/containers/create` errors there shouldn't be container left behind. Whether you used `--rm` on client side or not does not change that.

cpuguy83 commented:
OK, I see.

LGTM.
